src/clojure_sandbox/4clojure.clj

```diff
@@ -0,0 +1,605 @@
+
+;; 58. Write a function which allows you to create function compositions. The
+;;     parameter list should take a variable number of functions, and create a
+;;     function applies them from right-to-left.
+(defn comp* [& fns]
+  (fn [& args]
+    (loop [[current & remaining] (reverse fns)
+           result args]
+      (if (nil? current)
+        (first result)
+        (recur remaining (list (apply current result)))))))
+
+(comment ((comp* rest reverse) [1 2 3 4]))
+(comment ((comp* (partial + 3) second) [1 2 3 4]))
+
+;; 59. Take a set of functions and return a new function that takes a variable
+;;     number of arguments and returns a sequence containing the result of
+;;     applying each function left-to-right to the argument list.
+(defn juxt* [& fns]
+  (fn [& args]
+    (loop [[current & remaining] fns
+           result []]
+      (if (nil? current)
+        result
+        (recur remaining (conj result (apply current args)))))))
+
+(comment ((juxt* + max min) 2 3 5 1 6 4))
+(comment ((juxt* #(.toUpperCase %) count) "hello"))
+(comment ((juxt* :a :c :b) {:a 2 :b 4 :c 6 :d 8 :e 10}))
+
+;; 60. Write a function which behaves like reduce, but returns each intermediate
+;;     value of the reduction. Your function must accept either two or three
+;;     arguments, and the return sequence must be lazy.
+(defn reductions*
+  ([f coll] (reductions* f (first coll) (rest coll)))
+  ([f val coll] (lazy-seq
+                 (if (empty? coll)
+                   [val]
+                   (cons val (reductions* f (f val (first coll)) (rest coll)))))))
+
+(comment (take 5 (reductions* + (range))))
+(comment (reductions* conj [1] [2 3 4]))
+(comment (reductions* * 2 [3 4 5]))
+
+;; 61. Write a function which takes a vector of keys and a vector of values and
+;;     constructs a map from them.
+(defn zipmap* [keys values]
+  (loop [result {}
+         [current-key & remaining-keys] keys
+         [current-value & remaining-values] values]
+    (if (or (nil? current-key) (nil? current-value))
+      result
+      (recur (conj result {current-key current-value}) remaining-keys remaining-values))))
+
+(comment (zipmap* [:a :b :c] [1 2 3]))
+(comment (zipmap* [1 2 3 4] ["one" "two" "three"]))
+(comment (zipmap* [:foo :bar] ["foo" "bar" "baz"]))
+
+;; 62. Given a side-effect free function f and an initial value x write a
+;;     function which returns an infinite lazy sequence of x, (f x), (f (f x)),
+;;     (f (f (f x))), etc.
+(defn iterate* [f x]
+  (lazy-seq (cons x (iterate* f (apply f [x])))))
+
+(comment (take 5 (iterate* #(* 2 %) 1)))
+(comment (take 100 (iterate* inc 0)))
+(comment (take 9 (iterate* #(inc (mod % 3)) 1)))
+
+;; 63. Given a function f and a sequence s, write a function which returns a
+;;     map. The keys should be the values of f applied to each item in s. The
+;;     value at each key should be a vector of corresponding items in the order
+;;     they appear in s.
+(defn group-by* [f s]
+  (loop [result {}
```

Comment:
question: why not use reduce here

Reply:
Sorry, it didn't occur to me at that time. I changed it to use `reduce`.